File: app/src/main/java/com/cybersec/smartroute/engine/MockLocationEngine.java
package com.cybersec.smartroute.engine;

import android.content.Context;
import android.location.Criteria;
import android.location.Location;
import android.location.LocationManager;
import android.os.Build;
import android.os.SystemClock;
import android.util.Log;

/**
 * Real mock-GPS injector. Registers a test provider with the system
 * {@link LocationManager} and pushes coordinates that any other app reading
 * GPS will receive (when the device has Smart Route selected as the mock
 * location app under Developer options).
 */
public final class MockLocationEngine {

    private static final String TAG = "MockLocationEngine";
    public static final String PROVIDER = "smart_route_gps";

    private static volatile LocationManager locationManager;

    private MockLocationEngine() {
    }

    public static void attach(Context context) {
        if (locationManager == null) {
            synchronized (MockLocationEngine.class) {
                if (locationManager == null) {
                    locationManager = (LocationManager) context.getApplicationContext()
                            .getSystemService(Context.LOCATION_SERVICE);
                }
            }
        }
    }

    public static boolean initProvider() {
        LocationManager lm = locationManager;
        if (lm == null) return false;
        try {
            removeProvider();
            lm.addTestProvider(
                    PROVIDER,
                    /* requiresNetwork */ false,
                    /* requiresSatellite */ false,
                    /* requiresCell */ false,
                    /* hasMonetaryCost */ false,
                    /* supportsAltitude */ true,
                    /* supportsSpeed */ true,
                    /* supportsBearing */ true,
                    Criteria.POWER_LOW,
                    Criteria.ACCURACY_FINE);
            lm.setTestProviderEnabled(PROVIDER, true);
            return true;
        } catch (SecurityException se) {
            Log.w(TAG, "Mock provider registration denied: " + se.getMessage());
            return false;
        } catch (IllegalArgumentException iae) {
            try {
                lm.setTestProviderEnabled(PROVIDER, true);
                return true;
            } catch (Exception ex) {
                Log.w(TAG, "Provider re-enable failed: " + ex.getMessage());
                return false;
            }
        } catch (Exception ex) {
            Log.e(TAG, "initProvider failed", ex);
            return false;
        }
    }

    public static boolean setLocation(double lat, double lon) {
        return setLocation(lat, lon, 3f, 0d, 0f, 0f);
    }

    public static boolean setLocation(double lat, double lon, float accuracy,
                                      double altitude, float bearing, float speed) {
        LocationManager lm = locationManager;
        if (lm == null) return false;
        try {
            if (!lm.getAllProviders().contains(PROVIDER) || !lm.isProviderEnabled(PROVIDER)) {
                if (!initProvider()) return false;
            }
            Location location = new Location(PROVIDER);
            location.setLatitude(lat);
            location.setLongitude(lon);
            location.setAccuracy(accuracy);
            location.setAltitude(altitude);
            location.setBearing(bearing);
            location.setSpeed(speed);
            location.setTime(System.currentTimeMillis());
            location.setElapsedRealtimeNanos(SystemClock.elapsedRealtimeNanos());
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                location.setBearingAccuracyDegrees(0.5f);
                location.setSpeedAccuracyMetersPerSecond(0.5f);
                location.setVerticalAccuracyMeters(2f);
            }
            lm.setTestProviderLocation(PROVIDER, location);
            return true;
        } catch (SecurityException se) {
            Log.w(TAG, "setTestProviderLocation denied: " + se.getMessage());
            return false;
        } catch (Exception ex) {
            Log.e(TAG, "setLocation failed", ex);
            return false;
        }
    }

    public static void removeProvider() {
        LocationManager lm = locationManager;
        if (lm == null) return;
        try {
            if (lm.getAllProviders().contains(PROVIDER)) {
                lm.setTestProviderEnabled(PROVIDER, false);
                lm.removeTestProvider(PROVIDER);
            }
        } catch (Exception ignored) {
        }
    }
}

File: android/app/src/main/kotlin/com/cybersec/smartroute/MockLocationEngine.kt
package com.cybersec.smartroute

import android.content.Context
import android.location.Location
import android.location.LocationManager
import android.os.Build
import android.os.SystemClock

/**
 * Injects GPS fixes through Android's test location provider.
 * Used by the Flutter UI layer and the foreground service.
 */
object MockLocationEngine {
    const val PROVIDER = "smart_route_gps"

    private var locationManager: LocationManager? = null

    fun attach(context: Context) {
        if (locationManager == null) {
            locationManager =
                context.applicationContext.getSystemService(Context.LOCATION_SERVICE)
                    as LocationManager
        }
    }

    fun initProvider(): Boolean {
        val lm = locationManager ?: return false
        return try {
            removeProvider()
            lm.addTestProvider(
                PROVIDER,
                false,
                false,
                false,
                false,
                true,
                true,
                true,
                android.location.Criteria.POWER_LOW,
                android.location.Criteria.ACCURACY_FINE,
            )
            lm.setTestProviderEnabled(PROVIDER, true)
            true
        } catch (_: SecurityException) {
            false
        } catch (_: IllegalArgumentException) {
            try {
                lm.setTestProviderEnabled(PROVIDER, true)
                true
            } catch (_: Exception) {
                false
            }
        }
    }

    fun setLocation(
        lat: Double,
        lon: Double,
        accuracy: Float = 3f,
        altitude: Double = 0.0,
        bearing: Float = 0f,
        speed: Float = 0f,
    ): Boolean {
        val lm = locationManager ?: return false
        return try {
            if (!lm.allProviders.contains(PROVIDER) || !lm.isProviderEnabled(PROVIDER)) {
                if (!initProvider()) return false
            }
            val location = Location(PROVIDER).apply {
                latitude = lat
                longitude = lon
                this.accuracy = accuracy
                this.altitude = altitude
                this.bearing = bearing
                this.speed = speed
                time = System.currentTimeMillis()
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                    elapsedRealtimeNanos = SystemClock.elapsedRealtimeNanos()
                }
            }
            lm.setTestProviderLocation(PROVIDER, location)
            true
        } catch (_: SecurityException) {
            false
        }
    }

    fun removeProvider() {
        val lm = locationManager ?: return
        try {
            if (lm.allProviders.contains(PROVIDER)) {
                lm.setTestProviderEnabled(PROVIDER, false)
                lm.removeTestProvider(PROVIDER)
            }
        } catch (_: Exception) {
        }
    }
}
